### PR TITLE
feat: allow Callables for transport and channel init

### PR DESCRIPTION
**Pull request by daniel-sanche**
Our gapic libraries support passing in custom [transport](https://github.com/googleapis/python-bigtable/blob/7f747e10d5c4182d95327321872b4d78519de818/google/cloud/bigtable_v2/services/bigtable/client.py#L368) and [channel](https://github.com/googleapis/python-bigtable/blob/7f747e10d5c4182d95327321872b4d78519de818/google/cloud/bigtable_v2/services/bigtable/transports/grpc.py#L54) objects when initializing clients and transports respectively. 

In theory, this allows full customization of the transport stack to support uncommon or experimental use-cases, like channel pooling. In practice, customizing transports and channels in this way is messy, because the user needs to provide a set of parameters when constructing the transport or channel. These parameters depend on complex initialization logic (e.g., local credential parsing), which we already perform when a transport or channel is not provided. The user should not have to duplicate that code.

This change addresses the issue by allowing the user to provide a Callable in place of a custom transport or channel instance. The Callable will receive all the arguments that are already computed in the generator when the transport or channel is not provided, and is expected to return a usable instance (subclass of XXX or YYY).

We recommend that henceforth users generally provide such a callable rather than already instantiated transport or channel objects.

---

**Example**:

Instead of:
```
client_options = client_options_lib.ClientOptions()
api_endpoint, client_cert_source_func = self.get_mtls_endpoint_and_cert_source(client_options)
api_key_value = getattr(client_options, "api_key", None)

custom_transport = MyCustomTransport(
  credentials=None,
  credentials_file=client_options.credentials_file,
  host=api_endpoint,
  scopes=client_options.scopes,
  client_cert_source_for_mtls=client_cert_source_func,
  quota_project_id=client_options.quota_project_id,
  client_info=DEFAULT_CLIENT_INFO,
  always_use_jwt_access=True,
  api_audience=client_options.api_audience,
)
client = BigtableAsyncClient(transport=custom_transport)
```

We can do:
```
client = BigtableAsyncClient(transport=MyCustomTransport)
```

or:
```
client = BigtableAsyncClient(transport=lambda *args, **kwargs: MyCustomTransport(*args,  pool_size=5, **kwargs))
```

---

Let me know if you have any feedback. If we decide to go forward with his, I can add some tests and polish

### CODE REVIEW

**Comment by parthea** [comment]
Converting to draft as tests are failing. Please convert back to review once tests are passing.

**Comment by daniel-sanche** [comment]
> Converting to draft as tests are failing. Please convert back to review once tests are passing.

I have most of the tests passing, but I am having trouble interpreting the last "tests / integration". Any pointers?

**Comment by daniel-sanche** [comment]
tests are now passing

File: tests/integration/goldens/credentials/google/iam/credentials_v1/services/iam_credentials/client.py
Comment (vchudnov-g) [inline tests/integration/goldens/credentials/google/iam/credentials_v1/services/iam_credentials/client.py:385]: Here were invoking the callable with a specific set of arguments. The documentation above simply says "with the set of initialization arguments". Where are these arguments listed/described for the user? Maybe we should amend the parameters to say "the same set of initialization arguments as the provided transport classes [LINK]"

WDYT?

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2
Comment (vchudnov-g) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2:178]: nit: Here and in similar places below, I would consider adding a comment `# initialize with the provided callable or the default channel`. That would make it more immediately clear what's going on here.

**Review by vchudnov-g — COMMENTED** [review]
The functional changes appear good. I have one non-trivial in-code documentation suggestion.

Also, is this part of a greater initiative described in an internal doc? I couldn't find it in the one I was looking at. (Not that this needs lengthy documentation; it's straightforward enough that a one-liner would do, but it would be good to document why we need this.)

File: tests/integration/goldens/credentials/google/iam/credentials_v1/services/iam_credentials/client.py
Comment (daniel-sanche) [reply tests/integration/goldens/credentials/google/iam/credentials_v1/services/iam_credentials/client.py:385]: Yeah makes sense, although I'm not really sure what a proper link would look like in this case. I improved the docstrings, let me know if that looks better

**Comment by parthea** [comment]
Marking as draft as presubmits are failing

**Comment by daniel-sanche** [comment]
fixed presubmits

**Review by parthea — APPROVED** [review]

**Comment by parthea** [comment]
LGTM, but please wait for review from @vchudnov-g

**Comment by parthea** [comment]
@vchudnov-g Please can you take a look?

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2
Comment (vchudnov-g) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:450]: Can we hyperlink this somehow? Ditto elsewhere, and ditto for the channels.

**Review by vchudnov-g — APPROVED** [review]
Could you expand the PR description to include something like the following, if it is accurate? This is just to be very clear in the PR description and the commit message.

===BEGIN

... In practice, customizing `transport`s and `channel`s in this way is messy, because the user needs to provide a set of parameters when constructing the transport or channel. These parameters depend on  complex initialization logic (e.g., local credential parsing), which we already perform when a `transport` or `channel` is not provided. The user should not have to duplicate that code.

This change addresses the issue by allowing the user to provide a `Callable` in place of a custom `transport` or `channel` instance. The Callable will receive all the arguments that are already computed in the generator when the `transport` or `channel` is not provided, and is expected to return a usable instance (subclass of XXX or YYY).

We recommend that henceforth users generally provide such a callable rather than already instantiated `transport` or `channel` objects.

=== END

I wonder whether we should incorporate that last paragraph into code comments somehow. Maybe mention the callable is preferred? Maybe as a usage example in the `__init__` docstring?

**Pull request merged**
